cmd/protogen/main.go

```diff
@@ -0,0 +1,345 @@
+// The MIT License
+//
+// Copyright (c) 2022 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/format"
+	"go/parser"
+	"go/token"
+	"io/fs"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"go.temporal.io/api/internal/errgroup"
```

Comment:
Meh, probably not worth having/vendoring this as there is probably very little benefit to concurrently doing any of this. Just keep it simple. Surely not worth a whole new package of vendored code in this repo for such likely-unnecessary use. If you must have this, I would support making this cmd be its own `go.mod` with `golang.org/x/sync/errgroup` imported.

Reply:
I started by doing that but was unable to `go install` it from other repos, unfortunately. Perhaps I misconfigured something; you can see that in one of the earlier commits

Reply:
Hrmm, how did you install something from a fork before it was even merged? We know this works, because it's how people install our [`workflowcheck` tool](https://github.com/temporalio/sdk-go/tree/master/contrib/tools/workflowcheck) which is a child module of the primary repo.

Regardless, probably don't need a separate module or errgroup

Reply:
As for concurrency: I'm a fan of fast builds. I'll gather some simple timing metrics of the concurrent implementation and compare that against a serial one so we can decide whether it's worth the effort.

Reply:
The current concurrent implementation run against the server repo is around a full second faster per run.

Concurrent:
```
$ hyperfine --warmup 3 'make protoc'
Benchmark 1: make protoc
  Time (mean ± σ):      2.021 s ±  0.140 s    [User: 1.969 s, System: 1.768 s]
  Range (min … max):    1.816 s …  2.249 s    10 runs
```

Serial implementation:
```
$ hyperfine --warmup 3 'make protoc'
Benchmark 1: make protoc
  Time (mean ± σ):      2.955 s ±  0.100 s    [User: 1.966 s, System: 1.555 s]
  Range (min … max):    2.769 s …  3.108 s    10 runs
```

Reply:
I don't think you'd find anyone that isn't a fan of fast builds. I'm sure we could pour over the code and eek out some more wins if we were micro-optimizing, but I question the overall value, _especially_ since this is to replace the existing makefile that was probably serial.

Reply:
Sure. I'd prefer to optimize it further, but I'd rather have something maintainable in first. I've removed all concurrency with the latest commit